src/runtime/server/lib/oauth/oidc.ts

```diff
@@ -0,0 +1,326 @@
+  street_address?: string
+  /** City or locality component */
+  locality?: string
+  /** State, province, prefecture, or region component */
+  region?: string
+  /** Zip code or postal code component */
+  postal_code?: string
+  /** Country name component */
+  country?: string
+}
+
+interface OidcTokens {
+  access_token: string
+  scope: string
+  token_type: string
+}
+
+interface OIDCConfiguration {
+  authorization_endpoint: string
+  token_endpoint: string
+  userinfo_endpoint?: string
+}
+
+/**
+ * Event handler for generic OAuth using OIDC and PKCE.
+ */
+export function defineOAuthOidcEventHandler<TUser = OidcUser>({ config, onSuccess, onError }: OAuthConfig<OAuthOidcConfig, { user: TUser, tokens: OidcTokens }>) {
+  return eventHandler(async (event: H3Event) => {
+    config = defu(config, useRuntimeConfig(event).oauth?.oidc, {
+      scope: ['openid'],
+    } satisfies OAuthOidcConfig)
+
+    const query = getQuery<{ code?: string, error?: string, state?: string }>(event)
+
+    if (query.error) {
+      const error = createError({
+        statusCode: 401,
+        message: `OIDC login failed: ${query.error || 'Unknown error'}`,
+        data: query,
+      })
+      if (!onError) throw error
+      return onError(event, error)
+    }
+
+    if (!config.clientId || !config.openidConfig) {
+      return handleMissingConfiguration(event, 'oidc', ['clientId', 'openidConfig'], onError)
+    }
+
+    const oidcConfig = typeof config.openidConfig === 'string' ? await $fetch<OIDCConfiguration>(config.openidConfig) : config.openidConfig
+
+    const redirectURL = config.redirectURL || getOAuthRedirectURL(event)
+    const state = await handleState(event)
+
+    // if no client secret is provided, we will use PKCE so no client secret is needed
+    const verifier = !config.clientSecret ? await handlePkceVerifier(event) : undefined
+
+    if (!query.code) {
+      config.scope = config.scope || []
+
+      return sendRedirect(
+        event,
+        withQuery(oidcConfig.authorization_endpoint, {
+          client_id: config.clientId,
+          redirect_uri: redirectURL,
+          scope: config.scope.join(' '),
+          state,
+          response_type: 'code',
+          code_challenge: verifier?.code_challenge,
+          code_challenge_method: verifier?.code_challenge_method,
+          ...config.parameters?.authorization_endpoint,
+        }),
+      )
+    }
+
+    if (query.state !== state) {
+      return handleInvalidState(event, 'oidc', onError)
+    }
+
+    const tokens = await requestAccessToken<OidcTokens & { error?: unknown }>(oidcConfig.token_endpoint, {
+      body: {
+        grant_type: 'authorization_code',
+        client_id: config.clientId,
+        client_secret: config.clientSecret,
+        redirect_uri: redirectURL,
+        code: query.code,
+        code_verifier: verifier?.code_verifier,
```

Comment:
this part does not seem to omit itself when a clientSecret is provided, resulting in a 400 Bad Request error when fetching the tokens.

Reply:
In line 269 the verifier is set to undefined if a client secret is passed so its omitted in this case.

For me using a client secret works as expected 🤔  Which OIDC provider are you using?

Reply:
Thanks for testing! I updated the implementation to only include the PKCE query parameters if the client ID is unset.
Would you mind testing it again?

Reply:
Works as expected now!